torchrl/modules/functional_modules.py

```diff
@@ -0,0 +1,227 @@
+        model_copy = deepcopy(model)
+        param_tensordict = extract_weights(model_copy)
+        if disable_autograd_tracking:
+            param_tensordict.apply(lambda x: x.requires_grad_(False), inplace=True)
+        return FunctionalModule(model_copy), param_tensordict
+
+    def forward(self, params, *args, **kwargs):
+        # Temporarily load the state back onto self.stateless_model
+        old_state = _swap_state(
+            self.stateless_model, params, return_old_tensordict=_RESET_OLD_TENSORDICT
+        )
+        try:
+            return self.stateless_model(*args, **kwargs)
+        finally:
+            # Remove the loaded state on self.stateless_model
+            if _RESET_OLD_TENSORDICT:
+                _swap_state(self.stateless_model, old_state)
+
+
+class FunctionalModuleWithBuffers(nn.Module):
+    """
+    This is the callable object returned by :func:`make_functional`.
+    """
+
+    def __init__(self, stateless_model):
+        super(FunctionalModuleWithBuffers, self).__init__()
+        self.stateless_model = stateless_model
+
+    @staticmethod
+    def _create_from(model, disable_autograd_tracking=False):
+        # TODO: We don't need to copy the model to create a stateless copy
+        model_copy = deepcopy(model)
+        param_tensordict = extract_weights(model_copy)
+        buffers = extract_buffers(model_copy)
+        if buffers is None:
+            buffers = TensorDict(
+                {}, param_tensordict.batch_size, device=param_tensordict.device
+            )
+        if disable_autograd_tracking:
+            param_tensordict.apply(lambda x: x.requires_grad_(False), inplace=True)
+        return FunctionalModuleWithBuffers(model_copy), param_tensordict, buffers
+
+    def forward(self, params, buffers, *args, **kwargs):
+        # Temporarily load the state back onto self.stateless_model
+        old_state = _swap_state(
+            self.stateless_model, params, return_old_tensordict=_RESET_OLD_TENSORDICT
+        )
+        old_state_buffers = _swap_state(
+            self.stateless_model, buffers, return_old_tensordict=_RESET_OLD_TENSORDICT
+        )
+
+        try:
+            return self.stateless_model(*args, **kwargs)
+        finally:
+            # Remove the loaded state on self.stateless_model
+            if _RESET_OLD_TENSORDICT:
+                _swap_state(self.stateless_model, old_state)
+                _swap_state(self.stateless_model, old_state_buffers)
+
+
+def extract_weights(model):
+    tensordict = TensorDict({}, [])
+    for name, param in list(model.named_parameters(recurse=False)):
+        setattr(model, name, None)
+        tensordict[name] = param
+    for name, module in model.named_children():
+        module_tensordict = extract_weights(module)
+        if module_tensordict is not None:
+            tensordict[name] = module_tensordict
+    if len(tensordict.keys()):
+        return tensordict
+    else:
+        return None
```

Comment:
For TensorDict integration with `functorch.make_functional`, it looks like you've taught the API how to handle loading and unloading state from a TensorDict. Furthermore, make_functional returns a TensorDict instead of a list.

I like how we've changed make_functional to return a dictionary-like thing instead of a list. One question I have is if we should have make_functional return a TensorDict, a dictionary, or give users an option between those.

It depends on how much of a "PyTorch core" feature we think TensorDict is. As a litmus test: 
- Right now, given an nn.Module, one can get a state_dict from it. This state_dict is a regular dictionary. Do you think it should be a TensorDict?

Reply:
Let me be bold here: yes I think it should.
You can easily transform it to a tree structure:
TensorDict.unflatten_keys() will turn a flat tensordict in one that has a structure identical to you neural net. It's much easier to construct and to manipulate.

IMO it's an ideal structure for that, for transforms (eg when you want to transform a mask and an image in segmentation tasks etc), you can abstract away lots of stuff with it.

I think it does not have its place in TorchRL but right now that the best I can do with it. 
I think many usages people have of Raggedtensors from torchrec would also be simpler with tensordict (but I'm biased obviously).

@shagunsodhani suggested that we could do a separate repo under pytorch for it, I'm not against the idea. There's a workplace post on our group and tutorials on torchrl about it if you're interested.